src/libAtomVM/atom_table.c

```diff
@@ -0,0 +1,479 @@
+    table->last_node_group_avail--;
+    init_node(node, string, new_index);
+    insert_node_into_bucket(table, bucket_index, node);
+
+    return new_index;
+}
+
+static void do_rehash(struct AtomTable *table, int new_capacity)
+{
+    int new_size_bytes = sizeof(struct HNode *) * new_capacity;
+    struct HNode **new_buckets = realloc(table->buckets, new_size_bytes);
+    if (IS_NULL_PTR(new_buckets)) {
+        // Allocation failure can be ignored, the hash table will continue with the previous bucket
+        return;
+    }
+    memset(new_buckets, 0, new_size_bytes);
+    table->buckets = new_buckets;
+
+    struct HNodeGroup *group = table->first_node_group;
+
+    while (group) {
+        int group_count;
+        if (group == table->last_node_group) {
+            group_count = group->len - table->last_node_group_avail;
+        } else {
+            group_count = group->len;
+        }
+
+        for (int i = 0; i < group_count; i++) {
+            struct HNode *node = &group->nodes[i];
+            AtomString key = node->key;
+
+            unsigned long hash = sdbm_hash(key, atom_string_len(key));
+            unsigned long bucket_index = hash % table->capacity;
+
+            insert_node_into_bucket(table, bucket_index, node);
+        }
+
+        group = group->next;
+    }
+}
+
+static inline void maybe_rehash(struct AtomTable *table, int new_entries)
+{
+    int new_count = table->count + new_entries;
+    int threshold = ATOM_TABLE_THRESHOLD(table->capacity);
+    if (new_count > threshold) {
+        return;
+    }
+
+    int new_capacity = ATOM_TABLE_NEW_CAPACITY(new_count);
+    do_rehash(table, new_capacity);
+}
+
+long atom_table_ensure_atom(struct AtomTable *table, AtomString string, enum AtomTableCopyOpt opts)
+{
+    unsigned long hash = sdbm_hash(string, atom_string_len(string));
+    SMP_WRLOCK(table);
+    unsigned long bucket_index = hash % table->capacity;
+
+    struct HNode *node = get_node_from_bucket(table, bucket_index, string);
+    if (node) {
+        SMP_UNLOCK(table);
+        return node->index;
+    }
+    if (opts & AtomTableAlreadyExisting) {
+        SMP_UNLOCK(table);
+        return ATOM_TABLE_NOT_FOUND;
+    }
+
+    struct HNodeGroup *node_group = table->last_node_group;
+    if (!table->last_node_group_avail) {
+        node_group = new_node_group(table, DEFAULT_SIZE);
+        if (IS_NULL_PTR(node_group)) {
+            SMP_UNLOCK(table);
+            return ATOM_TABLE_ALLOC_FAIL;
+        }
+    }
+
+    AtomString maybe_copied = string;
+    if (opts & AtomTableCopyAtom) {
+        uint8_t len = *((uint8_t *) string);
+        uint8_t *buf = malloc(1 + len);
+        if (IS_NULL_PTR(buf)) {
+            SMP_UNLOCK(table);
+            return ATOM_TABLE_ALLOC_FAIL;
+        }
+        memcpy(buf, string, 1 + len);
+        maybe_copied = buf;
+    }
+
+    maybe_rehash(table, 1);
+
+    long new_index = insert_node(table, node_group, bucket_index, maybe_copied);
```

Comment:
I believe there is a bug here. bucket_index was computed before the capacity could have been incremented by maybe_rehash.

So I'd say we need more test.

Reply:
Fixed.